### ISSUE TITLE
Add GPU stats features

### ISSUE THREAD

**Issue by nicolargo** [giampaolo/psutil#526]
GPU are more and more used in scientific servers. It will be nice to have GPU stats features into PSUtil.

For examples of existing monitoring GPU software for Intel, NVidia or AMD GPU, see the post http://www.rkblog.rk.edu.pl/w/p/monitoring-amd-intel-and-nvidia-graphics-card-usage-under-linux/

Source (in C) already exist for Intel GPU Top:  http://anonscm.debian.org/cgit/pkg-xorg/app/intel-gpu-tools.git/tree/tools/intel_gpu_top.c

I will be a very nice feature asked byt Glances' users.

**Comment by giampaolo** [comment]
I'm willing to reopen this to investigate whether there are viable options to implement this at least for nvidia cards as it seems they are the most used in the scientifi community.

**Comment by Gerardwx** [comment]
NVIDIA already provides a library and has pypi package that provides python2 bindings, at least. https://developer.nvidia.com/nvidia-management-library-nvml

**Comment by ReenigneArcher** [comment]
Please add this! It would be a very nice addition to psutil.

Nvidia's official python module is [pynvml](https://pypi.org/project/pynvml/)

For AMD I found this module, [pyamdgpuinfo](https://pypi.org/project/pyamdgpuinfo/), but it is currently linux only.

Yet another library, but can only get very basic information is [gpu-info](https://github.com/Avlyssna/gpu-info) ... it's also on pypi, but no description there.

**Comment by DanielWicz** [comment]
Generally one wants some wrapper around nvidia-smi or rocm-smi (AMD). There's also intel's gpu (though less popular)

**Comment by XuehaiPan** [comment]
> NVIDIA already provides a library and has pypi package that provides python2 bindings, at least. [developer.nvidia.com/nvidia-management-library-nvml](https://developer.nvidia.com/nvidia-management-library-nvml)

Wrapping around `nvidia-smi` or `rocm-smi` might be the most feasible approach. The official NVIDIA Python binding [`nvidia-ml-py`](https://pypi.org/project/nvidia-ml-py) does not guarantee backward compatibility with old NVIDIA drivers. I reported compatibility concerns on the NVIDIA forum
 [[PyPI/nvidia-ml-py] Issue Reports for `nvidia-ml-py`](https://forums.developer.nvidia.com/t/pypi-nvidia-ml-py-issue-reports-for-nvidia-ml-py/196506).

See also:

- https://github.com/wookayin/gpustat/pull/107#issuecomment-893513321
- wookayin/gpustat#143
- XuehaiPan/nvitop#29
- XuehaiPan/nvitop#30
- XuehaiPan/nvitop#13

**Comment by giampaolo** [comment]
From your bug report:

> Backward compatibility between driver and binding versions.
> Since CUDA 11, the definition of nvmlProcessInfo_t adds two new fields gpuInstanceId and computeInstanceId.

[...]

> Another breaking change. nvidia-ml-py 11.515.0 (Jan 12, 2022) now even introduces v3 (nvmlDeviceGetComputeRunningProcesses_v3, etc.).

This is concerning [1]. If the C lib breaks compatibility so easily, psutil would probably have to use `#ifdef nvidia_version_x ... #else ...` clauses all over the place, and that may create problems with the binary wheels that we distribute on PYPI. The system compiling the wheel may have a certain nvidia-lib version supporting functionality "X", but the user installing the psutil wheel may not, and that usually results in `"X symbol not found"` error at import time. I faced a similar problem in https://github.com/giampaolo/psutil/pull/1879, which led me to rewrite  `prlimit()` functionality from C to ctypes for that reason. The fix was a literal "check for existence of X at run time instead of compilation time".

Also, right now we only depend on `apt-get install python3-dev` (Debian / Ubuntu) or `yum install python3-devel` (RedHat). Adding support for Nvidia GPUs means we'll have to also install `nvidialib-dev` / `nvidialib-devel` (or whatever they're called).  But not all Linux distros will provide a pre-compiled `nvidialib-dev` package, so we'll probably want logic in `setup.py` to make GPU functionality optional (aka not crash at compile time).

...and then there is Windows.

All of this to say that implementing this in pure C would probably be a nightmare, which is probably why they decided to use ctypes in pynvml: https://github.com/gpuopenanalytics/pynvml/blob/master/pynvml/nvml.py

In summary: if we'll ever want add GPU functionality in psutil we'll probably want to do the same and use ctypes.

[1] On a personal note: as a Linux user who's been dealing with Nvidia cards/driver issues for over a decade I'm not that surprised. ;)
